File: clients/python/mosaic_logger/logger.py
"""
Core Logger implementation for Mosaic Logger Python client
"""

import asyncio
from typing import List, Dict, Any, Optional, Union
from datetime import datetime
from .types import LogLevel, LogEvent, LogMessage, LogMetrics, ErrorInfo
from .context import LoggerContext, ContextManager
from .emoji import EmojiResolver
from .adapters.base import LogAdapter


class Logger:
    """Core logger implementation with emoji support and context management"""
    
    def __init__(
        self,
        service: str,
        environment: str,
        adapters: List[LogAdapter],
        emojis: bool = False,
        context: Optional[LoggerContext] = None,
        emoji_resolver: Optional[EmojiResolver] = None
    ):
        """
        Initialize logger
        
        Args:
            service: Service name for identification
            environment: Environment (dev, staging, prod)
            adapters: List of output adapters
            emojis: Whether to include emojis in log messages
            context: Base context for all log messages
            emoji_resolver: Custom emoji resolver
        """
        self.service = service
        self.environment = environment
        self.adapters = adapters
        self.emojis = emojis
        self.base_context = context or LoggerContext()
        self.emoji_resolver = emoji_resolver or EmojiResolver()
        
    async def log(
        self,
        level: LogLevel,
        message: str,
        event: Optional[LogEvent] = None,
        custom_event: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
        metrics: Optional[LogMetrics] = None,
        error: Optional[ErrorInfo] = None,
        tags: Optional[List[str]] = None,
        context: Optional[LoggerContext] = None
    ):
        """
        Core logging method
        
        Args:
            level: Log severity level
            message: Log message text
            event: Pre-defined event type
            custom_event: Custom event name for extensibility
            metadata: Additional structured data
            metrics: Performance and business metrics
            error: Error information
            tags: List of tags for categorization
            context: Context information for this log entry
        """
        # Merge contexts: base -> current -> provided
        current_context = ContextManager.get_context()
        final_context = self.base_context
        if current_context:
            final_context = final_context.merge(current_context)
        if context:
            final_context = final_context.merge(context)
            
        # Create log message
        log_message = LogMessage(
            timestamp=datetime.utcnow(),
            level=level,
            message=message,
            service=self.service,
            event=event,
            correlation_id=final_context.correlation_id,
            user_id=final_context.user_id,
            session_id=final_context.session_id,
            request_id=final_context.request_id,
            metadata=metadata,
            metrics=metrics,
            error=error,
            tags=tags
        )
        
        # Get emoji if enabled
        emoji = None
        if self.emojis:
            emoji = self.emoji_resolver.get_emoji(event, custom_event)
            
        # Format and send to all adapters
        tasks = []
        for adapter in self.adapters:
            formatted_message = adapter.format_message(log_message, self.emojis, emoji)
            task = adapter.write(log_message, formatted_message)
            tasks.append(task)
            
        if tasks:
            try:
                await asyncio.gather(*tasks, return_exceptions=True)
            except Exception:
                # Ignore adapter errors to prevent logging failures from breaking application
                pass
                
    async def debug(self, message: str, **kwargs):
        """Log debug message"""
        await self.log(LogLevel.DEBUG, message, **kwargs)
        
    async def info(self, message: str, **kwargs):
        """Log info message"""
        await self.log(LogLevel.INFO, message, **kwargs)
        
    async def warn(self, message: str, **kwargs):
        """Log warning message"""
        await self.log(LogLevel.WARN, message, **kwargs)
        
    async def error(self, message: str, **kwargs):
        """Log error message"""
        await self.log(LogLevel.ERROR, message, **kwargs)
        
    # Synchronous convenience methods
    def debug_sync(self, message: str, **kwargs):
        """Synchronous debug logging"""
        asyncio.create_task(self.debug(message, **kwargs))
        
    def info_sync(self, message: str, **kwargs):
        """Synchronous info logging"""
        asyncio.create_task(self.info(message, **kwargs))
        
    def warn_sync(self, message: str, **kwargs):
        """Synchronous warning logging"""
        asyncio.create_task(self.warn(message, **kwargs))
        
    def error_sync(self, message: str, **kwargs):
        """Synchronous error logging"""
        asyncio.create_task(self.error(message, **kwargs))
        
    async def close(self):
        """Close all adapters and cleanup resources"""
        tasks = [adapter.close() for adapter in self.adapters]
        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)
            
    def with_context(self, **kwargs) -> 'Logger':
        """Create a logger with additional context"""
        new_context = self.base_context
        
        # Update context with provided values
        context_update = LoggerContext(
            correlation_id=kwargs.get('correlation_id'),
            user_id=kwargs.get('user_id'),
            session_id=kwargs.get('session_id'),
            request_id=kwargs.get('request_id'),
            trace_id=kwargs.get('trace_id'),
            span_id=kwargs.get('span_id'),
            custom_context={k: v for k, v in kwargs.items() 
                          if k not in ['correlation_id', 'user_id', 'session_id', 'request_id', 'trace_id', 'span_id']}
        )
        
        merged_context = new_context.merge(context_update)
        
        return Logger(
            service=self.service,
            environment=self.environment,
            adapters=self.adapters,
            emojis=self.emojis,
            context=merged_context,
            emoji_resolver=self.emoji_resolver
        )

File: clients/typescript/src/logger.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Core Logger class for Mosaic Logger
import type {
  ErrorContext,
  ErrorDetails,
  Logger as ILogger,
  LogAdapter,
  LogEntry,
  LogLevel,
  LoggingContext,
} from './types';
import type { EmojiResolver } from './emoji';
import { ContextManager, createCorrelationId } from './context';

/**
 * Core Logger implementation with emoji support and context management
 */
export class Logger implements ILogger {
  private service: string;
  private environment: string;
  private emojiResolver: EmojiResolver;
  private contextManager: ContextManager;
  private adapters: LogAdapter[];
  private baseContext: Partial<LoggingContext>;

  constructor(
    service: string,
    environment: string,
    emojiResolver: EmojiResolver,
    adapters: LogAdapter[],
    baseContext: Partial<LoggingContext> = {}
  ) {
    this.service = service;
    this.environment = environment;
    this.emojiResolver = emojiResolver;
    this.adapters = adapters;
    this.baseContext = baseContext;
    this.contextManager = new ContextManager();

    // Set initial context
    if (Object.keys(baseContext).length > 0) {
      this.contextManager.setContext(baseContext);
    }
  }

  /**
   * Log a trace level message
   */
  trace(message: string, data?: Partial<LogEntry>): void {
    this.log('TRACE', message, data);
  }

  /**
   * Log a debug level message
   */
  debug(message: string, data?: Partial<LogEntry>): void {
    this.log('DEBUG', message, data);
  }

  /**
   * Log an info level message
   */
  info(message: string, data?: Partial<LogEntry>): void {
    this.log('INFO', message, data);
  }

  /**
   * Log a warn level message
   */
  warn(message: string, data?: Partial<LogEntry>): void {
    this.log('WARN', message, data);
  }

  /**
   * Log an error level message
   */
  error(message: string, data?: Partial<LogEntry>): void {
    this.log('ERROR', message, data);
  }

  /**
   * Log a fatal level message
   */
  fatal(message: string, data?: Partial<LogEntry>): void {
    this.log('FATAL', message, data);
  }

  /**
   * Log a message at the specified level
   */
  log(level: LogLevel, message: string, data?: Partial<LogEntry>): void {
    const entry = this.createLogEntry(level, message, data);
    this.writeToAdapters(entry);
  }

  /**
   * Create a child logger with additional context
   */
  child(context: Partial<LoggingContext>): Logger {
    const mergedContext = { ...this.baseContext, ...context };
    const childLogger = new Logger(
      this.service,
      this.environment,
      this.emojiResolver,
      this.adapters,
      mergedContext
    );
    return childLogger;
  }

  /**
   * Flush all adapters
   */
  async flush(): Promise<void> {
    const flushPromises = this.adapters
      .filter(
        (adapter): adapter is LogAdapter & { flush: NonNullable<LogAdapter['flush']> } =>
          adapter.flush !== undefined
      )
      .map((adapter) => adapter.flush());

    await Promise.all(flushPromises);
  }

  /**
   * Validate and sanitize error objects to ensure they conform to ErrorDetails interface
   */
  private sanitizeError(error: unknown): ErrorDetails | undefined {
    if (error === null || error === undefined) {
      return undefined;
    }

    // Handle JavaScript Error instances
    if (error instanceof Error) {
      return {
        type: error.constructor.name,
        message: error.message,
        stackTrace: error.stack,
      };
    }

    // Handle string errors
    if (typeof error === 'string') {
      return {
        type: 'Unknown',
        message: error,
      };
    }

    // Handle object-like errors
    if (typeof error === 'object') {
      const errorObj = error as Record<string, unknown>;

      // Check if it has the minimum required fields for ErrorDetails
      if (typeof errorObj.type === 'string' && typeof errorObj.message === 'string') {
        const sanitized: ErrorDetails = {
          type: errorObj.type,
          message: errorObj.message,
        };

        // Add optional fields if present and valid
        if (typeof errorObj.stackTrace === 'string') {
          sanitized.stackTrace = errorObj.stackTrace;
        }
        if (typeof errorObj.code === 'string') {
          sanitized.code = errorObj.code;
        }
        if (typeof errorObj.context === 'object' && errorObj.context !== null) {
          sanitized.context = errorObj.context as ErrorContext;
        }

        return sanitized;
      }

      // If it's an object but doesn't have required fields, ignore it
      return undefined;
    }

    // For any other type, ignore it
    return undefined;
  }
  /**
   * Create a complete log entry
   */
  private createLogEntry(level: LogLevel, message: string, data?: Partial<LogEntry>): LogEntry {
    const timestamp = new Date();
    const logId = this.generateLogId();
    const currentContext = this.contextManager.getContext();

    // Merge contexts: base -> current -> data.context
    const context: LoggingContext = {
      ...this.baseContext,
      ...currentContext,
      ...(data?.context ?? {}),
    };

    // Ensure we have a correlation ID
    context.correlationId ??= createCorrelationId();

    // Sanitize error object
    const sanitizedError = data?.error ? this.sanitizeError(data.error) : undefined;

    const entry: LogEntry = {
      logId,
      timestamp,
      level,
      message,
      service: this.service,
      environment: this.environment,
      includeEmoji: this.emojiResolver.isEnabled(),
      context,
      ...(data?.event && { event: data.event }),
      ...(data?.metadata && { metadata: data.metadata }),
      ...(data?.metrics && { metrics: data.metrics }),
      ...(sanitizedError && { error: sanitizedError }),
    };

    return entry;
  }

  /**
   * Generate a unique log ID
   */
  private generateLogId(): string {
    // Generate log ID in format: log_<32 hex chars>
    const chars = '0123456789abcdef';
    let result = 'log_';
    for (let i = 0; i < 32; i++) {
      result += chars[Math.floor(Math.random() * 16)];
    }
    return result;
  }

  /**
   * Write log entry to all adapters
   */
  private writeToAdapters(entry: LogEntry): void {
    this.adapters.forEach((adapter) => {
      try {
        const result = adapter.write(entry);
        if (result && typeof result === 'object' && 'then' in result) {
          void Promise.resolve(result).catch((error) => {
            // Don't let adapter errors break logging
            // eslint-disable-next-line no-console
            console.error('Logger adapter error:', error);
          });
        }
      } catch (error) {
        // Don't let adapter errors break logging
        // eslint-disable-next-line no-console
        console.error('Logger adapter error:', error);
      }
    });
  }

  /**
   * Get current context
   */
  getContext(): LoggingContext {
    return this.contextManager.getContext();
  }

  /**
   * Set context for this logger
   */
  setContext(context: Partial<LoggingContext>): void {
    this.contextManager.setContext(context);
  }

  /**
   * Get service name
   */
  getService(): string {
    return this.service;
  }

  /**
   * Get environment
   */
  getEnvironment(): string {
    return this.environment;
  }

  /**
   * Check if emojis are enabled
   */
  isEmojisEnabled(): boolean {
    return this.emojiResolver.isEnabled();
  }

  /**
   * Enable or disable emojis
   */
  setEmojisEnabled(enabled: boolean): void {
    this.emojiResolver.setEnabled(enabled);
  }

  /**
   * Get the emoji resolver
   */
  getEmojiResolver(): EmojiResolver {
    return this.emojiResolver;
  }

  /**
   * Close the logger and clean up resources
   */
  async close(): Promise<void> {
    // Flush all adapters first
    await this.flush();

    // Close adapters that support it
    const closePromises = this.adapters
      .filter(
        (adapter): adapter is LogAdapter & { close: NonNullable<LogAdapter['close']> } =>
          adapter.close !== undefined
      )
      .map((adapter) => adapter.close());

    await Promise.all(closePromises);
  }
}
